### PR TITLE
Rdm 213 Overriding security classification in callback

### PR DESCRIPTION
**Pull request by andrewfolga**
**Before creating a pull request make sure that:**
- [x] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-213

### Change description ###
Overriding security classification in callback

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/model/callbacks/CallbackResponse.java [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.databind.JsonNode;
+import io.swagger.annotations.ApiModelProperty;
+import uk.gov.hmcts.ccd.data.casedetails.SecurityClassification;
 
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
+//@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/callbacks/CallbackResponse.java]: leftover?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createcase/SubmitCaseTransaction.java [hunk 2]
```diff
@@ -78,6 +78,14 @@ public CaseDetails submitCase(Event event,
          */
         callbackInvoker.invokeAboutToSubmitCallback(eventTrigger, null, newCaseDetails, caseType, ignoreWarning);
 
+        final CaseDetails savedCaseDetails = saveAuditEventForCaseDetails(event, caseType, idamUser, eventTrigger, newCaseDetails);
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createcase/SubmitCaseTransaction.java]: This is extra code cleanup refactor not related to the story.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 3]
```diff
@@ -55,26 +53,27 @@ public SecurityClassificationService(@Qualifier(CachedUserRepository.QUALIFIER)
...
+                                                       MAPPER.convertValue(caseDetails.getDataClassification(), JsonNode.class),
+                                                       securityClassification);
                     caseDetails.setData(MAPPER.convertValue(data, STRING_JSON_MAP));
                     return cd;
-                });
+                }));
     }
 
-    public List<AuditEvent> apply(String jurisdictionId, List<AuditEvent> events) {
+    public List<AuditEvent> applyClassification(String jurisdictionId, List<AuditEvent> events) {
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: This is extra method rename refactor not related to the story.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 4]
```diff
@@ -116,35 +115,36 @@ private JsonNode filterNestedObject(JsonNode data, JsonNode dataClassification,
         while (dataIterator.hasNext()) {
             Map.Entry<String, JsonNode> dataElement = dataIterator.next();
             String dataElementKey = dataElement.getKey();
-            JsonNode dataClassificationParent = dataClassification.get(dataElementKey);
-            if (isNull(dataClassificationParent)) {
+            JsonNode dataClassificationElement = dataClassification.get(dataElementKey);
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: This is extra variable name refactor not related to the story.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 5]
```diff
@@ -183,46 +185,14 @@ private void filterObject(SecurityClassification userClassification, Iterator<Ma
         }
     }
 
-    private boolean isNull(JsonNode... jsonNodes) {
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: This is extracted into new SecurityClassificationUtils class.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 6]
```diff
@@ -183,46 +185,14 @@ private void filterObject(SecurityClassification userClassification, Iterator<Ma
...
-        return newArrayList(jsonNodes).stream().anyMatch(Objects::isNull);
-    }
-
-    private JsonNode getDataClassificationForData(JsonNode data, Iterator<JsonNode> dataIterator) {
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: Same as above.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 7]
```diff
@@ -183,46 +185,14 @@ private void filterObject(SecurityClassification userClassification, Iterator<Ma
...
-    }
-
     private void filterSimpleField(SecurityClassification userClassification, Iterator iterator, JsonNode dataClassificationValue) {
         Optional<SecurityClassification> securityClassification = getSecurityClassification(dataClassificationValue);
         if (!securityClassification.isPresent() || !userClassification.higherOrEqualTo(securityClassification.get())) {
             iterator.remove();
         }
     }
 
-    private Optional<SecurityClassification> getSecurityClassification(JsonNode dataNode) {
```
Comment (andrewfolga) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: This is extracted into new SecurityClassificationUtils class.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java [hunk 8]
```diff
@@ -487,6 +496,41 @@ void shouldFilterAllFieldsForCaseWithMultiSelectListTypes() throws IOException {
             );
         }
 
+        @Test
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java]: New tests that were missing for SecurityClassificationService to increase branch coverage.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java [hunk 9]
```diff
@@ -864,10 +929,77 @@ void shouldFilterFieldsWithinCollectionItems() throws IOException {
...
+        // TODO Target implementation, see RDM-1204
+        @DisplayName("should filter out collection item if no item classification exists")
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java]: New tests that were missing for SecurityClassificationService to increase branch coverage.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java [hunk 10]
```diff
@@ -918,6 +1050,57 @@ void shouldFilterOutCollectionItemsHigherClassification() throws IOException {
             );
         }
 
+        @Test
+        // TODO Target implementation, see RDM-1204
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationServiceTest.java]: New tests that were missing for SecurityClassificationService to increase branch coverage.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/stdapi/CallbackInvokerTest.java [hunk 11]
```diff
@@ -150,13 +160,42 @@ void sendCallbackAndGetState() {
             assertThat(response.get(), is(expectedState));
         }
 
+        @Test
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/stdapi/CallbackInvokerTest.java]: New test that was missing for CallbackInvoker to increase branch coverage.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 12]
```diff
@@ -55,26 +53,27 @@ public SecurityClassificationService(@Qualifier(CachedUserRepository.QUALIFIER)
                              .max(comparingInt(SecurityClassification::getRank));
     }
 
-    public Optional<CaseDetails> apply(CaseDetails caseDetails) {
+    public Optional<CaseDetails> applyClassification(CaseDetails caseDetails) {
         Optional<SecurityClassification> userClassificationOpt = getUserClassification(caseDetails.getJurisdiction());
         Optional<CaseDetails> result = Optional.of(caseDetails);
 
-        return result.filter(caseIfHasClassificationEqualOrLowerThan(userClassificationOpt))
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: not sure why this change but if it's equivalent to the old version I think the old one is simpler to understand. 
Doing a filter first followed by a map is simpler than nesting the filtering inside and doing a flat map

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseTypeService.java [hunk 13]
```diff
@@ -56,6 +59,7 @@ public void validateData(final Map<String, JsonNode> data,
                              final CaseType caseType) {
         final List<ValidationResult> dataValidationResults = caseDataValidator.validate(data, caseType.getCaseFields());
         if (!dataValidationResults.isEmpty()) {
+            LOG.warn("There have been validation errors={}", dataValidationResults);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseTypeService.java]: what kind of validation errors will be logging as WARN? things like invalid email?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java [hunk 14]
```diff
@@ -183,46 +185,14 @@ private void filterObject(SecurityClassification userClassification, Iterator<Ma
...
-            LOG.error("Unable to parse security classification for {}", dataNode, e);
-            return Optional.empty();
-        }
-        return Optional.of(securityClassification);
-    }
-
-    private Predicate<CaseDetails> caseIfHasClassificationEqualOrLowerThan(Optional<SecurityClassification> userClassificationOpt) {
-        return cd -> userClassificationOpt.map(uc -> uc.higherOrEqualTo(cd.getSecurityClassification())).orElse(false);
+    private boolean isNull(JsonNode... jsonNodes) {
+        return newArrayList(jsonNodes).stream().anyMatch(Objects::isNull);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: is anyMatch correct here? or should be allMatch?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseTypeService.java (hunk 13 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseTypeService.java]: I guess yes all sorts of field validation errors.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java (hunk 14 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/common/SecurityClassificationService.java]: Yeah should be allMatch.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityValidationServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,974 @@
...
+        securityValidationService = new SecurityValidationService();
+    }
+
+    @Nested
+    @DisplayName("Validate data classification case")
+    class ValidateDataClassificationCase {
+
+        @Test
+        @DisplayName("should fail if invalid classification level for case")
+        void shouldFailIfInvalidClassificationLevelForCase() {
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityValidationServiceTest.java]: Add a happy path test.

File: src/test/java/uk/gov/hmcts/ccd/endpoint/std/CallbackTest.java [hunk 16]
```diff
@@ -9,7 +9,6 @@
 import org.hamcrest.Matchers;
```
Comment (andrewfolga) [inline src/test/java/uk/gov/hmcts/ccd/endpoint/std/CallbackTest.java]: Add a component test that verifies/asserts the entire logic works together.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityValidationServiceTest.java (hunk 15 above)
Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/SecurityValidationServiceTest.java]: Done.
